README
MSIM 1.5.0 -- Machine Simulator
===============================

To compile and install:
./configure && make install

This codebase has been derived from the MSIM 1.4.0 [1] (which originally
stood for MIPS Simulator). It currently simulates the MIPS R4000 CPU.

[1] http://d3s.mff.cuni.cz/~holub/sw/msim/


CHANGELOG
2018-01-12 Martin Decky <martin@decky.cz>
	Version 1.5.0
	* new development branch derived from the
	  latest MSIM 1.5.0
	* the acronym now means Machine Simulator
	  in anticipation of future changes

2016-03-01 Martin Decky <decky@d3s.mff.cuni.cz>
	Version 1.4.0
	* improve command line processing code
	* code consolidation
	* port fixes from the 1.3 branch

2010-10-05 Martin Decky <decky@d3s.mff.cuni.cz>
	Version 1.3.8
	* add support for DESTDIR into Makefile.in
	  (thx Vojtech Horky)
	* fix TLBWI implementation (thx Jan Zaloha
	  and Vlastimil Babka)
	* resurrect GDB support, better device management
	  and various other improvements by Tomas
	  Martinec

2009-11-02 Martin Decky <decky@d3s.mff.cuni.cz>
	Version 1.3.7.1
	* add Gentoo ebuild and RPM spec file

2009-02-02 Martin Decky <decky@d3s.mff.cuni.cz>
	Version 1.3.7
	* official reference manual in doc/reference.html
	* minor string changes

2009-01-19 Martin Decky <decky@d3s.mff.cuni.cz>
	Version 1.3.6
	* fix compilation in Cygwin
	* use EPC of previous branch instruction if TLB exception
	  on code fetch happens inside a branch delay slot
	  (thx Ondrej Cerny)
	* fix memory access alignment

2008-10-31 Martin Decky <decky@d3s.mff.cuni.cz>
	Version 1.3.5.1
	* fix console input in MinGW

2008-10-26 Martin Decky <decky@d3s.mff.cuni.cz>
	Version 1.3.5
	* initial support for building in MinGW
	* code and data breakpoints
	* support WATCH exception (thx Lubomir Bulej)

2008-09-24 Martin Decky <decky@d3s.mff.cuni.cz>
	Version 1.3.4
	* source cleanup
	* simplify build system
	* split sources into multiple directories
	* add GPL license

2008-06-05 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.3.4rc2
	* LL-SC tracking fix (thx Jiri Svoboda)

2007-12-17 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.3.4rc1
	* dorder sychdown command fix (thx Andrej Krutak)
	* ddisk and mem clean-up (thx Milan Burda)
	* dtime now reads usec as well
	* minor performance improvements
	* badvaddr not set for all exceptions, fixed

2007-11-30 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.3.3
	* two-register dorder
	* error messages fix
	* cpu number fix

2007-11-24 Viliam Holub <holub@d3s.mff.cuni.cz>
	* dorder has two registers now - enable/disable interrupt pending

2007-11-23 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.3.2.3
	* dorder infinite loop fix (thx David Matousek)

2007-11-21 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.3.2.2
	* applied patch by Martin Decky, several small fixes

2007-03-11 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.3.2.1
	* missing stdio.h in input.c fixed
	* MULT fix (thx Ondrej Palkovsky <ondrap@penguin.cz>)
	* u_int64_t -> uint_64 fix
	* dprintf renamed to mprintf to avoid conflicts with libc (Martin
	  Decky <decky@d3s.mff.cuni.cz>
	* minor fixes
	* goto has been renamed to continue
	* autogen.sh added

2006-10-09 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.3.1
	* dprintf* functions have been renamed to mprintf* to avoid conflicts
	  with libc on some systems (thx to Martin Decky)

2005-12-30 Ondrej Palkovsky <ondrap@penguin.cz>
	* keybord keycode register is cleared after read operation; this
	  allows to read new keycodes even with disabled interrupts

2005-10-18 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.3
	* several small fixes
	* error messages cleanup

2005-10-10 Jakub Jermar <jermar@itbs.cz>
	* TLB Modification exception fix

2005-09-20 Ondrej Palkovsky <ondrap@penguin.cz>
	* add 4kc support

2005-09-13 Viliam Holub <holub@d3s.mff.cuni.cz>
	* LWL, LWR, SWL, SWR byte counting fix by Ondrej Palkovsky

2005-03-03 Viliam Holub <holub@d3s.mff.cuni.cz>
	* MTC0 fix (thx Ondrej Palkovsky)
	* applied fixes by Lubo
	* code polishing
	* polishing of check RQ

2005-01-08 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.2.18
	* minor changes
	* the first version of TAB completion
	* all system and device command are now defined via structures
	* malloc and strdup safe
	* ERET fix, LBU fix, epc rewrite fix (thx Matej Pivoluska,
	  Martin Horvath, Jakub Kotrla)
	* code polishing

2004-03-23 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.2.12
	* device command are now generic
	* generic context help

2003-12-18 Viliam Holub <holub@d3s.mff.cuni.cz>
	* code clean-up

2003-12-17 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.2.3
	* cycle counter size increased to long long
	* non-POSIX terminal attributes are now checked
	* code cleanup (comments, read/write functions)
	* Changelog cleanup
	* (dcpu.c) md command fix
	* endians checked via autoconf

2003-12-01 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.2.2
	* TLB dirty mask fix (read-only pages were writable)

2003-11-28 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.2.1
	* MTLO and MTHI fix

2003-11-24 Viliam Holub <holub@d3s.mff.cuni.cz>
	Version 1.2
	* debug messages fix
	* added dtime


src/text.c
/*
 * Copyright (c) 2002-2004 Viliam Holub
 * All rights reserved.
 *
 * Distributed under the terms of GPL.
 *
 *
 *  String constants
 *
 */

#include <errno.h>
#include <string.h>
#include "../config.h"
#include "text.h"
#include "main.h"

/*
 * The most frequently messages for user.
 * Device implementation should prefer these.
 */
const char *const txt_file_read_err     = "Could not read file";
const char *const txt_filename_expected = "File name expected";
const char *const txt_file_create_err   = "Could not create file";
const char *const txt_file_write_err    = "Could not write to file";
const char *const txt_file_map_fail     = "File map fail";
const char *const txt_devname_expected  = "Device name expected";
const char *const txt_duplicate_devname = "Duplicate device name";
const char *const txt_devaddr_expected  = "Device address expected";
const char *const txt_devaddr_error     = "Device address error (4-byte alignment expected)";
const char *const txt_no_more_parms     = "No more parameters allowed";
const char *const txt_not_en_mem        = "Not enough memory for device inicialization";
const char *const txt_intnum_expected   = "Interrupt number expected";
const char *const txt_intnum_range      = "Interrupt number out of range 0..6";
const char *const txt_cmd_expected      = "Command expected";
const char *const txt_unknown_cmd       = "Unknown command";

const char *txt_exc[] = {
	/* 0 */
	"Interrupt",
	"TLB modification",
	"TLB (load or instruction fetch)",
	"TLB (store)",
	"Address error (load or instruction fetch)",
	"Address error (store)",
	"Bus error (instruction fetch)",
	"Bus error (data reference: load or store)",
	/* 8 */
	"Syscall",
	"Breakpoint",
	"Reserved instruction",
	"Coprocessor Unusable",
	"Arithmetic Overflow",
	"Trap",
	"Virtual Coherency instruction",
	"Floating-Point",
	/* 16 */
	"Exception 16",
	"Exception 17",
	"Exception 18",
	"Exception 19",
	"Exception 20",
	"Exception 21",
	"Exception 22",
	"Reference to WatchHi/WatchLo address",
	/* 24 */
	"Exception 24",
	"Exception 25",
	"Exception 26",
	"Exception 27",
	"Exception 28",
	"Exception 29",
	"Exception 30",
	"Virtual Coherency data"
};

const char txt_version[] =
	"MSIM version " VERSION "\nCopyright (c) 2000-2018 Viliam Holub, Martin Decky\n";

const char txt_help[] =
	"  -V, --version            display version info\n"
	"  -c, --config=file_name   configuration file name\n"
	"  -i, --interactive        enter interactive mode\n"
	"  -t, --trace              enter trace mode\n"
	"  -g, --remote-gdb=port    enter gdb mode\n"
	"  -n, --non-deterministic  enable non-deterministic behaviour\n";

const char hexchar[] = "0123456789abcdef";
